Fixes ssm-agent for 21.05

joedevivo opened the pull request:
###### Motivation for this change
ssm-agent wasn't starting on ec2 with a 21.05 AMI/

###### Things done
It turns out that ssm-agent expects a few files in `/etc/amazon/ssm`:
* amazon-ssm-agent.json.template
* seelog.xml.template

Unfortunately, the `/etc/amazon/ssm` is hardcoded; however, we've got `substituteInPlace` so I swapped in an ENV var `SSM_AGENT_ETC` and added that to the wrapper script.

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of ~all~ binary files (usually in `./result/bin/`)
    - tested the wrapper, which launches the other bins
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Most importantly, I built an AMI with this package and removed the /etc/amazon/ssm entries in my `configuration.nix`.

The instance is reachable via ssm :tada:

joedevivo committed:
ssm-agent: copy files required for agent startup

cole-h commented on pkgs/applications/networking/cluster/ssm-agent/default.nix:
```diff
@@ -63,6 +63,9 @@ buildGoPackage rec {
         --replace '"script"' '"${util-linux}/bin/script"'
 
+    substituteInPlace agent/appconfig/constants_unix.go \
+        --replace '"/etc/amazon/ssm/"' 'os.Getenv("SSM_AGENT_ETC") + "/"'
```
I think this should work, and seems better than patching in a nonstandard env var.

```suggestion
        --replace '"/etc/amazon/ssm/"' '"${placeholder "out"}/etc/amazon/ssm/"'
```

joedevivo replied:
I went with an ENV var in case someone wanted to override the location with their own configuration file. Is there another way to preserve that functionality? I initially wanted to go the way you're suggesting

cole-h replied:
Is there a reason one would want to change this config?

If so, this should probably be a patch to look in `$out/etc/amazon/ssm` when `SSM_AGENT_ETC` isn't set. By wrapping with `--set`, the override wouldn't work anyways (it would always take the value of the `--set` env var, AFAIK).

joedevivo replied:
Maybe I'm overthinking it. I'll stick with the default config. I'll make all the suggested changes tomorrow. Thanks for the feedback.

cole-h replied:
For what it's worth, you raise a good point. But as-is, our two changes are extremely similar

cole-h commented on pkgs/applications/networking/cluster/ssm-agent/default.nix:
```diff
@@ -96,8 +99,14 @@ buildGoPackage rec {
   '';
 
+  postInstall = ''
+    mkdir -p $out/etc
+    cp ${src}/amazon-ssm-agent.json.template $out/etc/.
+    cp ${src}/seelog_unix.xml $out/seelog.xml.template
```
Might as well mimic the upstream layout-ish.

```suggestion
    mkdir -p $out/etc/amazon/ssm
    cp amazon-ssm-agent.json.template $out/etc/amazon/ssm/amazon-ssm-agent.json.template
    cp seelog_unix.xml $out/etc/amazon/ssm/seelog.xml.template
```

joedevivo replied:
When I remove `${src}/`, I get this:

```
cp: cannot stat 'amazon-ssm-agent.json.template': No such file or directory
```

I had to use the prefix `go/src/${goPackagePath}/` to find it.

cole-h replied:
Ah, right. I forgot about that.

cole-h commented on pkgs/applications/networking/cluster/ssm-agent/default.nix:
```diff
@@ -96,8 +99,14 @@ buildGoPackage rec {
+
   postFixup = ''
-    wrapProgram $out/bin/amazon-ssm-agent --prefix PATH : ${bashInteractive}/bin
+    wrapProgram $out/bin/amazon-ssm-agent --set SSM_AGENT_ETC $out/etc --prefix PATH : ${bashInteractive}/bin
```
```suggestion
    wrapProgram $out/bin/amazon-ssm-agent --prefix PATH : ${bashInteractive}/bin
```

cole-h commented:
And also, please change the commit message to align with our contributing document. Concretely, this would look like `ssm-agent: copy files required for ec2` or something.

cole-h commented on pkgs/applications/networking/cluster/ssm-agent/default.nix:
```diff
@@ -96,6 +99,21 @@ buildGoPackage rec {
+  # module could use this template and substitute a different log level.
+  postInstall = ''
+    mkdir -p $out/etc/amazon/ssm
+    echo "$(pwd)"
```
```suggestion
```

joedevivo commented:
This should be ready to go. Let me know if it needs anything else.

cole-h reviewed (changes requested):
Let me start by saying: I appreciate your effort in solving this issue.

Doing some more thinking (and talking with somebody who has an opinion on this), I'm thinking this *should* actually go into `/etc` using the module, despite your efforts to not have to touch the module 😬

My main reasoning is actually related to your first attempt using an env var -- somebody somewhere down the line might actually want to change that configuration. With the current solution, they'd have to modify the derivation to copy their configuration into `$out/etc/....`. If the module were to handle the `/etc` stuff, then all they'd have to do is `environment.etc."/etc/.....".text = ""` or `.source = ""`.

So, my suggestion amounts to the following:

* in the package itself, copy these template configurations to `$out/share/ssm-agent/...`
* in the module, add an option for these configurations and have them be symlinked into place using `environment.etc` (they should default to the template files, IMO; but they should also be nullable, e.g. `nullOr`)

Sorry it took so long to get back to you (and sorry for asking for even more effort). If you think this is too much for you to take on, I'll gladly do so (though it might be a little bit before I can). Just let me know.

joedevivo commented:
No worries. I wasn't actively avoiding the module, but I was developing the fix on an Ubuntu machine, and it occurred to me that the package as is won't work on any distro without being able to find those templates. I think your proposed change would work for NixOS, but the package won't work on Ubuntu out of the box.

We can either say this is for NixOS only and rework the PR how you suggest, or leave it as is and say it'll work on any distro, but there's no overriding the configuration.

There could be a third option, but here's where I first say that I'm no expert and am just kind of guessing this sounds possible...

In the nixos module, we could override the package to not replace the `/etc/amazon/ssm` path and copy those default templates via environment.etc, with config options to override those templates. If I'm missing something here, feel free to call it out.

Which path do you think we should take here?

grahamc commented:
> In the nixos module, we could override the package to not replace the /etc/amazon/ssm path and copy those default templates via environment.etc

This sounds like a pretty good middle ground to me.

grahamc commented:
@joedevivo I was the person cole-h was referring to before -- would you go ahead and take the middle ground of linking files from the package to /etc in the NixOS module?

joedevivo commented:
Sure thing. Might be a day or two before I can get to it.

joedevivo committed:
nixos/ssm-agent: conf files written to /etc

ssm-agent expects files in /etc/amazon/ssm. The pkg substitutes a location in
the nix store for those default files, but if we ever want to adjust this
configuration on NixOS, we'd need the ability to modify that file.

This change to the nixos module writes copies of the default files from the nix
store to /etc/amazon/ssm. Future versions can add config, but right now this
would allow users to at least write out a text value to
environment.etc."amazon/ssm/amazon-ssm-agent.json".text to provide
their own config.

joedevivo commented:
Verified this works as expected on EC2. Everything works as the original PR did, which the additional verification of the presence of `/etc/amazon/ssm/amazon-ssm-agent.json.template` and `/etc/amazon/ssm/seelog.xml`

cole-h commented on nixos/modules/services/misc/ssm-agent.nix:
```diff
@@ -22,8 +22,8 @@ in {
       description = "The SSM agent package to use";
-      default = pkgs.ssm-agent;
-      defaultText = "pkgs.ssm-agent";
+      default = pkgs.callPackage ../../../../pkgs/applications/networking/cluster/ssm-agent { overrideEtc = false; };
+      defaultText = "pkgs.callPackage ../../../../pkgs/applications/networking/cluster/ssm-agent { overrideEtc = false; }";
```
```suggestion
      default = pkgs.ssm-agent.override { overrideEtc = false; };
      defaultText = "pkgs.ssm-agent.override { overrideEtc = false; }";
```

cole-h commented on nixos/modules/services/misc/ssm-agent.nix:
```diff
@@ -62,5 +64,10 @@ in {
+
+    environment.etc."amazon/ssm/seelog.xml".source = "${cfg.package}/seelog.xml.template";
+
+    environment.etc."amazon/ssm/amazon-ssm-agent.json.template".source =  "${cfg.package}/etc/amazon/ssm/amazon-ssm-agent.json.template";
```
```suggestion
    environment.etc."amazon/ssm/amazon-ssm-agent.json".source =  "${cfg.package}/etc/amazon/ssm/amazon-ssm-agent.json.template";
```

cole-h commented on pkgs/applications/networking/cluster/ssm-agent/default.nix:
```diff
@@ -63,6 +64,11 @@ buildGoPackage rec {
     substituteInPlace agent/session/shell/shell_unix.go \
         --replace '"script"' '"${util-linux}/bin/script"'
 
+    if [ ! -z "${toString overrideEtc}" ]; then
+      substituteInPlace agent/appconfig/constants_unix.go \
+        --replace '"/etc/amazon/ssm/"' '"${placeholder "out"}/etc/amazon/ssm/"'
+    fi
+
     echo "${version}" > VERSION
   '';
```
```suggestion
    echo "${version}" > VERSION
  '' + lib.optionalString overrideEtc ''
    substituteInPlace agent/appconfig/constants_unix.go \
      --replace '"/etc/amazon/ssm/"' '"${placeholder "out"}/etc/amazon/ssm/"'
  '';
```

cole-h reviewed (changes requested):
I like this approach much better, thanks for tackling it! A few minor things that should be adjusted.

joedevivo commented:
Should be good now

cole-h reviewed (approved):
Looks good indeed!

Just to confirm: you've tried running the most recent incarnation of this pull to make sure it works? I don't think any of my suggestions would have broken it, but better safe than sorry.
